File: src/main/java/fr/skyost/serialkey/listener/PadlockFinderListener.java
package fr.skyost.serialkey.listener;

import org.bukkit.ChatColor;
import org.bukkit.Location;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.block.Action;
import org.bukkit.event.player.PlayerInteractEvent;
import org.bukkit.inventory.ItemStack;

import fr.skyost.serialkey.SerialKey;

/**
 * A listener that allows to listen padlock finder related events.
 */

public class PadlockFinderListener extends SerialKeyListener {

	/**
	 * Creates a new padlock finder listener instance.
	 *
	 * @param plugin The plugin.
	 */

	public PadlockFinderListener(final SerialKey plugin) {
		super(plugin);
	}

	@EventHandler(priority = EventPriority.LOWEST)
	private void onPlayerInteract(final PlayerInteractEvent event) {
		if(event.isCancelled() || event.getAction() != Action.RIGHT_CLICK_AIR && event.getAction() != Action.RIGHT_CLICK_BLOCK) {
			return;
		}

		final ItemStack item = event.getItem();
		if(!api.isUsedPadlockFinder(item)) {
			return;
		}

		final Player player = event.getPlayer();
		final Location spawn = player.getWorld().getSpawnLocation();
		if(player.getCompassTarget().equals(spawn)) {
			try {
				player.setCompassTarget(api.extractLocation(item));
				player.sendMessage(plugin.getPluginMessages().message4);
			}
			catch(final Exception ex) {
				ex.printStackTrace();
				plugin.sendMessage(player, ChatColor.RED + ex.getClass().getName());
			}
		}
		else {
			player.setCompassTarget(spawn);
			player.sendMessage(plugin.getPluginMessages().message5);
		}
		event.setCancelled(true);
	}
	
}

File: core/src/main/java/fr/skyost/serialkey/core/listener/PadlockFinderListener.kt
package fr.skyost.serialkey.core.listener;

import fr.skyost.serialkey.core.SerialKeyPlugin;
import fr.skyost.serialkey.core.object.SerialKeyLocation;
import fr.skyost.serialkey.core.object.SerialKeyPerson;

import java.util.function.Consumer;

/**
 * A listener that allows to listen padlock finder related events.
 */

public class PadlockFinderListener<I, L> extends SerialKeyListener<I, L> {

	/**
	 * Creates a new padlock finder listener instance.
	 *
	 * @param plugin The plugin.
	 */

	public PadlockFinderListener(final SerialKeyPlugin<I, L> plugin) {
		super(plugin);
	}

	/**
	 * Triggered when a player right clicks.
	 *
	 * @param item The item.
	 * @param player The involved player.
	 * @param spawn The spawn location.
	 * @param currentCompassesTarget The current compasses target.
	 * @param setCompassesTarget Function that allows to change the current compasses target.
	 * @param cancelEvent The runnable that cancels the event.
	 */

	protected void onPlayerRightClick(final I item, final SerialKeyPerson player, final SerialKeyLocation spawn, final SerialKeyLocation currentCompassesTarget, final Consumer<SerialKeyLocation> setCompassesTarget, final Runnable cancelEvent) {
		if(!itemManager.isUsedPadlockFinder(item)) {
			return;
		}

		final SerialKeyLocation padlockLocation = unlocker.getLocations(item).get(0);
		if(currentCompassesTarget.equals(spawn)) {
			setCompassesTarget.accept(padlockLocation);
			plugin.sendMessage(player, plugin.getPluginMessages().getPadlockFinderEnabledMessage());
		}
		else {
			setCompassesTarget.accept(spawn);
			plugin.sendMessage(player, plugin.getPluginMessages().getPadlockFinderDisabledMessage());
		}

		cancelEvent.run();
	}

}